src/core/net/ip6.cpp

```diff
@@ -1199,9 +1235,17 @@ otError Ip6::HandleDatagram(Message &aMessage, Netif *aNetif, const void *aLinkM
         }
 #endif
 
-        IgnoreError(ProcessReceiveCallback(aMessage, messageInfo, nextHeader, aFromNcpHost, Message::kCopyToUse));
+        error = ProcessReceiveCallback(aMessage, messageInfo, nextHeader, aFromNcpHost, Message::kCopyToUse);
+
+        if ((error == OT_ERROR_NONE || error == OT_ERROR_NO_ROUTE) && forward &&
+            !ShouldForwardToThread(messageInfo, aFromNcpHost))
+        {
+            forward = false;
+        }
```

Comment:
```suggestion
        forward = forward && ShouldForwardToThread(messageInfo, aFromNcpHost);
```
Let's skip checking errors? I think retrying won't help in any case.

Reply:
It helps when failed for insufficient memory.